File: pocketpilot-backend/models/transaction.py
from datetime import datetime, timezone
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class TransactionType(str, Enum):
    INCOME = "income"
    EXPENSE = "expense"


class TransactionCategory(str, Enum):
    FOOD = "food"
    TRANSPORT = "transport"
    HOUSING = "housing"
    ENTERTAINMENT = "entertainment"
    EDUCATION = "education"
    GIFT = "gift"
    ONE_TIME = "one_time"
    OTHER = "other"


class TransactionClass(str, Enum):
    """How a transaction affects the user's monthly spending PLAN.

    Orthogonal to `type` (income/expense) and `category` (food/transport/
    etc). Answers: "should this move the needle on today's safe-to-spend?"
    """

    FIXED_COMMITMENT = "fixed_commitment"  # SIP, EMI, subscription, rent
    DISCRETIONARY = "discretionary"  # everyday spend, counts against daily allowance
    TRANSFER_INTERNAL = "transfer_internal"  # self-to-self, own accounts/wallets
    REFUND = "refund"  # reverses an earlier discretionary spend
    INCOME_REGULAR = "income_regular"  # salary/allowance/pocket money — funds the plan
    ONE_TIME_EXCEPTION = "one_time_exception"  # emergency/irregular, excluded from daily math
    UNCLASSIFIED = "unclassified"  # needs manual review


class ClassificationSource(str, Enum):
    RULE = "rule"  # decided by the rule-based classifier with high confidence
    USER = "user"  # user manually confirmed/corrected it
    PENDING = "pending"  # awaiting user review


class TransactionBase(BaseModel):
    amount: float = Field(..., gt=0)
    type: TransactionType = TransactionType.EXPENSE
    category: TransactionCategory = TransactionCategory.OTHER
    description: Optional[str] = Field(None, max_length=500)
    merchant: Optional[str] = Field(None, max_length=200)
    source: str = Field(default="manual")
    date: Optional[datetime] = Field(default_factory=lambda: datetime.now(timezone.utc))
    txn_class: TransactionClass = TransactionClass.DISCRETIONARY
    classification_source: ClassificationSource = ClassificationSource.USER
    classification_confidence: float = Field(default=1.0, ge=0, le=1)
    # Which of the user's registered bank accounts this transaction came
    # from/went to. Optional since older transactions predate multi-account
    # support, and manual entries may not specify one.
    account_id: Optional[str] = None
    # Original SMS text, when this transaction came from SMS parsing.
    # Surfaced in the transaction detail view so the user can see exactly
    # what the bank actually said, not just our parsed interpretation.
    raw_sms: Optional[str] = None
    # Which specific rule/keyword the classifier matched to reach its
    # decision (e.g. "discretionary_kw:swiggy") — surfaced on the detail
    # screen as "why was this classified this way", so a wrong guess is
    # obvious and explainable rather than a silent black box.
    classification_rule: Optional[str] = None
    # True when this transaction pushed that day's discretionary spend
    # above the flat daily_limit and the user hasn't yet decided how to
    # absorb the overage (savings / reduce future days / hybrid / treat as
    # an emergency exception). See routers/budget.py overage endpoints.
    overage_pending: bool = False
    overage_resolved_at: Optional[datetime] = None


class TransactionCreate(TransactionBase):
    pass


class TransactionUpdate(BaseModel):
    amount: Optional[float] = Field(None, gt=0)
    type: Optional[TransactionType] = None
    category: Optional[TransactionCategory] = None
    description: Optional[str] = Field(None, max_length=500)
    date: Optional[datetime] = None
    txn_class: Optional[TransactionClass] = None
    classification_source: Optional[ClassificationSource] = None
    classification_confidence: Optional[float] = Field(None, ge=0, le=1)
    account_id: Optional[str] = None


class TransactionResponse(TransactionBase):
    id: str
    user_id: str
    created_at: datetime
    updated_at: datetime

File: pocketpilot_app/lib/models/transaction.dart
enum TransactionType { income, expense }

enum TransactionCategory {
  food,
  transport,
  housing,
  entertainment,
  education,
  gift,
  oneTime,
  other
}

/// How a transaction affects the user's monthly spending PLAN.
/// Orthogonal to [TransactionType] and [TransactionCategory].
enum TransactionClass {
  fixedCommitment,
  discretionary,
  transferInternal,
  refund,
  incomeRegular,
  oneTimeException,
  unclassified,
}

enum ClassificationSource { rule, user, pending }

TransactionClass _transactionClassFromJson(String? value) {
  switch (value) {
    case 'fixed_commitment':
      return TransactionClass.fixedCommitment;
    case 'transfer_internal':
      return TransactionClass.transferInternal;
    case 'refund':
      return TransactionClass.refund;
    case 'income_regular':
      return TransactionClass.incomeRegular;
    case 'one_time_exception':
      return TransactionClass.oneTimeException;
    case 'unclassified':
      return TransactionClass.unclassified;
    default:
      return TransactionClass.discretionary;
  }
}

String _transactionClassToJson(TransactionClass value) {
  switch (value) {
    case TransactionClass.fixedCommitment:
      return 'fixed_commitment';
    case TransactionClass.discretionary:
      return 'discretionary';
    case TransactionClass.transferInternal:
      return 'transfer_internal';
    case TransactionClass.refund:
      return 'refund';
    case TransactionClass.incomeRegular:
      return 'income_regular';
    case TransactionClass.oneTimeException:
      return 'one_time_exception';
    case TransactionClass.unclassified:
      return 'unclassified';
  }
}

extension TransactionClassLabel on TransactionClass {
  String get label => switch (this) {
        TransactionClass.fixedCommitment => 'Fixed commitment',
        TransactionClass.discretionary => 'Discretionary spend',
        TransactionClass.transferInternal => 'Internal transfer',
        TransactionClass.refund => 'Refund',
        TransactionClass.incomeRegular => 'Income',
        TransactionClass.oneTimeException => 'One-time exception',
        TransactionClass.unclassified => 'Needs review',
      };

  /// Public wire-format serializer, used by ApiService when submitting a
  /// user's classification decision back to the backend.
  String get wireValue => _transactionClassToJson(this);
}

ClassificationSource _classificationSourceFromJson(String? value) {
  switch (value) {
    case 'rule':
      return ClassificationSource.rule;
    case 'pending':
      return ClassificationSource.pending;
    default:
      return ClassificationSource.user;
  }
}

TransactionType _transactionTypeFromJson(String value) =>
    value == 'income' ? TransactionType.income : TransactionType.expense;

String _transactionTypeToJson(TransactionType value) =>
    value == TransactionType.income ? 'income' : 'expense';

TransactionCategory _transactionCategoryFromJson(String value) {
  switch (value) {
    case 'food':
      return TransactionCategory.food;
    case 'transport':
      return TransactionCategory.transport;
    case 'housing':
      return TransactionCategory.housing;
    case 'entertainment':
      return TransactionCategory.entertainment;
    case 'education':
      return TransactionCategory.education;
    case 'gift':
      return TransactionCategory.gift;
    case 'one_time':
      return TransactionCategory.oneTime;
    default:
      return TransactionCategory.other;
  }
}

String _transactionCategoryToJson(TransactionCategory value) {
  switch (value) {
    case TransactionCategory.food:
      return 'food';
    case TransactionCategory.transport:
      return 'transport';
    case TransactionCategory.housing:
      return 'housing';
    case TransactionCategory.entertainment:
      return 'entertainment';
    case TransactionCategory.education:
      return 'education';
    case TransactionCategory.gift:
      return 'gift';
    case TransactionCategory.oneTime:
      return 'one_time';
    case TransactionCategory.other:
      return 'other';
  }
}

extension TransactionCategoryLabel on TransactionCategory {
  String get label => switch (this) {
        TransactionCategory.oneTime => 'one time',
        _ => name,
      };
}

class Transaction {
  final String id;
  final double amount;
  final TransactionType type;
  final TransactionCategory category;
  final String? description;
  final String? merchant;
  final String source;
  final DateTime date;
  final DateTime createdAt;
  final DateTime updatedAt;
  final TransactionClass txnClass;
  final ClassificationSource classificationSource;
  final double classificationConfidence;
  final String? accountId;
  final String? rawSms;
  final String? classificationRule;

  Transaction({
    required this.id,
    required this.amount,
    required this.type,
    required this.category,
    this.description,
    this.merchant,
    required this.source,
    required this.date,
    required this.createdAt,
    required this.updatedAt,
    this.txnClass = TransactionClass.discretionary,
    this.classificationSource = ClassificationSource.user,
    this.classificationConfidence = 1.0,
    this.accountId,
    this.rawSms,
    this.classificationRule,
  });

  bool get needsReview => classificationSource == ClassificationSource.pending;

  /// Human-readable explanation of why the classifier reached its
  /// decision, derived from the raw rule code (e.g. "discretionary_kw:
  /// swiggy"). Falls back to a generic note for manual entries or when no
  /// rule was recorded (older transactions predate this field).
  String get classificationExplanation {
    final rule = classificationRule;
    if (classificationSource == ClassificationSource.user) {
      return 'You confirmed this classification.';
    }
    if (rule == null) {
      return 'No detailed reason recorded for this transaction.';
    }
    final parts = rule.split(':');
    final kind = parts[0];
    final matched = parts.length > 1 ? parts[1] : null;

    switch (kind) {
      case 'autopay_name':
        return 'Matched your autopay named "$matched".';
      case 'self_account':
        return 'Matched your registered account "$matched".';
      case 'income_kw':
        return 'Detected the income keyword "$matched" in the SMS.';
      case 'refund_kw':
        return 'Detected the refund keyword "$matched" in the SMS.';
      case 'fixed_kw':
        return 'Detected the keyword "$matched", suggesting a recurring bill or subscription.';
      case 'transfer_kw':
        return 'Detected the keyword "$matched", suggesting a transfer between your own accounts.';
      case 'exception_kw':
        return 'Detected the keyword "$matched", suggesting an emergency or one-time expense.';
      case 'discretionary_kw':
        return 'Matched merchant keyword "$matched".';
      case 'no_text_income_default':
        return 'No SMS text available — defaulted to income since you marked it a credit.';
      case 'no_text_expense_default':
        return 'No SMS text available — defaulted to discretionary spend.';
      case 'fallback_income_unclassified':
      case 'fallback_expense_unclassified':
        return 'No confident match found — flagged for your review.';
      default:
        return 'Classified using rule: $rule';
    }
  }

  factory Transaction.fromJson(Map<String, dynamic> json) => Transaction(
        id: json['id'],
        amount: (json['amount'] as num).toDouble(),
        type: _transactionTypeFromJson(json['type']),
        category: _transactionCategoryFromJson(json['category']),
        description: json['description'],
        merchant: json['merchant'],
        source: json['source'],
        date: DateTime.parse(json['date']),
        createdAt: DateTime.parse(json['created_at']),
        updatedAt: DateTime.parse(json['updated_at']),
        txnClass: _transactionClassFromJson(json['txn_class']),
        classificationSource:
            _classificationSourceFromJson(json['classification_source']),
        classificationConfidence:
            (json['classification_confidence'] as num?)?.toDouble() ?? 1.0,
        accountId: json['account_id'],
        rawSms: json['raw_sms'],
        classificationRule: json['classification_rule'],
      );
}

class TransactionCreate {
  final double amount;
  final TransactionType type;
  final TransactionCategory category;
  final String? description;
  final String? merchant;
  final String source;
  final DateTime date;

  TransactionCreate({
    required this.amount,
    this.type = TransactionType.expense,
    this.category = TransactionCategory.other,
    this.description,
    this.merchant,
    this.source = 'manual',
    DateTime? date,
  }) : date = date ?? DateTime.now();

  Map<String, dynamic> toJson() => {
        'amount': amount,
        'type': _transactionTypeToJson(type),
        'category': _transactionCategoryToJson(category),
        'description': description,
        'merchant': merchant,
        'source': source,
        'date': date.toIso8601String(),
      };
}
